### PR TITLE
Pushbutton Handler

### PR DESCRIPTION
**Pull request by murgui**
Pushbutton handler, it is missing to test. However I would like to do it with the interface PCB as it is HW related and I guess it will be ready to go tomorrow. 

At the moment you can check the nice looking code ( ͡° ͜ʖ ͡°)

### CODE REVIEW

**Review by NicolaiFransen — APPROVED** [review]
Looks nice. I like how you used the structpointer again.

File: Software/Development/SemesterProject/App/pushbutttonManager.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+    for (structPointer = initialMemoryPosition; structPointer < finalMemoryPosition; structPointer++)
+    {
+        pushEventQualificationFSM(structPointer);
+    }
+}
+
+void pushEventQualificationFSM(PushButton *pushbutton)
+{
+    switch(pushbutton->FSMState)
```
Comment (murgui) [inline Software/Development/SemesterProject/App/pushbutttonManager.c:65]: Missing break; in every case

**Review by murgui — COMMENTED** [review]
To be changed

File: Software/Development/SemesterProject/App/pushbutttonManager.c [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+int torqueReferenceHasBeenPressed(void)
+{
+    int buttonStateValue = pushButtonList.torqueReference.ButtonState;
+    pushButtonList.torqueReference.ButtonState = 0;
+    return buttonStateValue;
+}
+
+int antiSlipHasBeenPressed(void)
```
Comment (murgui) [inline Software/Development/SemesterProject/App/pushbutttonManager.c]: substitute by speedRefDecreaseHasBeenPressed

File: Software/Development/SemesterProject/App/pushbutttonManager.c [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+{
+    return pushbutton->functionPointer();
+}
+
+int pinIsLow(PushButton *pushbutton)
+{
+    return !(pushbutton->functionPointer());
+}
+
+int regenerativeBrakingHasBeenPressed(void)
```
Comment (murgui) [inline Software/Development/SemesterProject/App/pushbutttonManager.c]: speedRefIncreaseHasBeenPressed

File: Software/Development/SemesterProject/App/pushbutttonManager.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ *      This implies risks in case that two software components use the same
+ *      button, but that will never be allowed.
+ *
+ */
+
+#include "pushButtonManager.h"
+
+static struct pushButtonListTag
+{
+    PushButton regenerativeBraking;
```
Comment (murgui) [inline Software/Development/SemesterProject/App/pushbutttonManager.c]: referenceSource
referenceType
speedIncrease
speedDecrease

File: Software/Development/SemesterProject/App/pushbutttonManager.c [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+int cruiseControlHasBeenPressed(void)
+{
+    int buttonStateValue = pushButtonList.cruiseControl.ButtonState;
+    pushButtonList.cruiseControl.ButtonState = 0;
+    return buttonStateValue;
+}
+
+int torqueReferenceHasBeenPressed(void)
```
Comment (AitorTeran) [inline Software/Development/SemesterProject/App/pushbutttonManager.c]: maybe better referenceSource?

File: Software/Development/SemesterProject/App/pushbutttonManager.c [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+int regenerativeBrakingHasBeenPressed(void)
+{
+    int buttonStateValue = pushButtonList.regenerativeBraking.ButtonState;
+    pushButtonList.regenerativeBraking.ButtonState = 0;
+    return buttonStateValue;
+}
+
+int cruiseControlHasBeenPressed(void)
```
Comment (AitorTeran) [inline Software/Development/SemesterProject/App/pushbutttonManager.c]: maybe better referenceType?
Also, this function will have to change it's value under circunstances different from button being pressed. For example when the driver brakes while it's active, the reference type has to change from speed to torque. Same for when driver accelerates momentarely or other cases. 
How are this situations going to be handled? maybe better changing the name to referenceTypeHasChanged? 
Maybe this is done somewhere else, idk but its something to keep in mind. Also we have to take a look at priorities and what to do under every situation.

**Review by AitorTeran — COMMENTED** [review]
Looking good!

File: Software/Development/SemesterProject/App/pushbutttonManager.c (hunk 5 above)
Comment (murgui) [reply Software/Development/SemesterProject/App/pushbutttonManager.c]: you got it👌

File: Software/Development/SemesterProject/App/pushbutttonManager.c (hunk 6 above)
Comment (murgui) [reply Software/Development/SemesterProject/App/pushbutttonManager.c]: reference type ✔️

priorities and doing this kind of situation analysis you describe is performed in reference handler. You are completely right that we definitely have to look very carefully into it.

**Comment by murgui** [comment]
I just commited the changes. The pull request will be a draft until I get to test in HW. It is missing to schedule the module.

**Pull request merged**